---
title: "Accessing Genome Data"
output:
  pdf_document: default
  html_notebook: default
---

reading a vcf file: 

```{bash}
pwd
```


```{r}
library(vcfR)

genome_table <- read.vcfR('imp_TERRA_RIL_SNP.vcf')
```

 accessing the fix, meta, and gt files: 

This gives the data for the reference DNA and puts it into a fll table: be careful with accessing the gt file since it is so large

```{r}
reference = genome_table@fix
View(reference)
gt_file = genome_table@gt
meta_file = genome_table@meta

```

The meta file gives you the meta data of the file: 

```{r}
meta_file = genome_table@meta
meta_file
```

To access a single genome (there are 304), access a column of the gt file (genome data starts at column 2). The genome is presented in the a|b format, where $a,b \in \{ 0,1 \}$.  The way to read this is comparing to the reference.  The reference gives you the column of reference or dominant alleles, but also a table of the alternate.  a sequence with 0|0 means that the entry has the dominant on both sides of the Helix, 0|1 and 1|0 means a dominant on one side, alternate on the other, and 1|1 means alterante alleles on both sides. To get the genome keys, simply take the 0th row. This gives you the "format" key, the rest of the keys, and then the "reference genome".  You can also access genomes with specific keys:

```{r}
table1 = gt_file[,2] #access a single genome, pick numbers between 2 and 305
key_names = gt_file[0,] # access the names of the file as a matrix
attr(gt_file,"dimnames") # access the names of files as a list of character strings
PI_tables=gt_file[,startsWith(colnames(gt_file),"PI")] # creates a table of genomes whose IDs begin with PI


```

Let's now establish a distance function.  We want to se how genomic distance correlates with different phenomic traits.  Start with with Hamming Distance: 

```{r}
hamming <- function(x,y) {
  if(length(x)!=length(y)){
    print('error, vectors not same size!')
  }else{
    result <- length(x[x!=y])
    
    return(result)
  }
}

hamming(gt_file[,3],gt_file[,10])

hamming_mag <- function(x) {
  result <- length(x[x!="0|0"])
  return(result)
}
distances <- vector()
for (i in 2:305){
  distances[i-1] <-hamming_mag(gt_file[,i])
}
hist(distances,breaks = 20)
plot(1:304,distances)
hamming_mag(gt_file[,2])
min(distances)
max(distances)

order(distances)
length(gt_file[,1])
```
Now Let's do something a bit more involved.  We know that 00 differs from 01 in a way distinct from 11, so let's record that. Let's say that |00-01| and |00-10| = 1, and |00-11| = 2, while 01 and 10 are considered the same, hence |01-10| is 0. From there, we can define the following distances: 

```{r}
#The following is an allele distance table which will define allele differences
names <-c('0|0','0|1','1|0','1|1')
allele_dif_table <- matrix(c(0,1,1,2,1,0,0,1,1,0,0,1,2,1,1,0),
                           nrow = 4,ncol = 4, dimnames = list(names, names))

#turns genomes into vectors with entries 0,1,2, where '0|0'-> 0; '0|1','1|0' ->1; '1|1'-> 2. 
tonumericform <- function(a){
  if (class(a)=="numeric") {
    b<-matrix(nrow = length(gt_file[,1]),ncol=2)
    b[,]   <-0       
    return(b)
    }
  else{
    vec<-matrix(nrow = length(a),ncol=2)
    vec[,1]<-as.numeric(substr(a, 1,1))
    vec[,2]<-as.numeric(substr(a, 3,3))
  return(vec)
} 
}
#cityblock distance: to get a cityblock magnitude, set a2 to '0|0'.  This is similar to the l_1 distance of a vector
cityblock<- function(a1,a2){

  vec1<-tonumericform(a1)
  vec2<-tonumericform(a2)
  dif<- length(vec1[vec1[,1]!=vec2[,1],1])+length(vec1[vec1[,2]!=vec2[,2],1])
  return(dif)
}
print(hamming(gt_file[,2],gt_file[,3]))
print(cityblock(gt_file[,2],gt_file[,3]))

#Euclidean distance: to get a Euclidean Magnitude, set a2 to '0|0''
euclidean<- function(a1,a2){
  vec1<-tonumericform(a1)
  vec2<-tonumericform(a2)
  dif <- norm(vec1-vec2,"2")
}

c<-print(euclidean(gt_file[,2],gt_file[,3]))

city_distances <- vector()
for (i in 2:305){
  city_distances[i-1] <-cityblock(gt_file[,i],0)
  i
}

euclid_distances <- vector()
for (i in 2:305){
  euclid_distances[i-1] <-euclidean(gt_file[,i],0)
  i
}

hist(city_distances,breaks = 25)
plot(1:304,city_distances)
hist(euclid_distances)
plot(1:304,euclid_distances)
```


Now we have more than one kind of distance and set of magnitudes.   Let's set up a table of the different kinds of tables:
```{r}
#distance_table <- matrix(c(distances,city_distances,euclid_distances),nrow = 304, ncol = 3)
#rownames(distance_table)<- colnames(gt_file)[2:305]
#colnames(distance_table)<- c('hamming','cityblock','euclidean')
#View(distance_table)

class(gt_file[,2])
b<-tonumericform(gt_file[,2])
a<-tonumericform(gt_file[,4])
b
class(b-a)
norm(b-a,"1")


```

```{r}
m=matrix(nrow=3,ncol=2)
m[1,]<-c(1,2)
m[2,]<-c(1,3)
m[3,]<-c(0,1)

n=matrix(nrow=3,ncol=2)
n[1,]<-c(1,4)
n[2,]<-c(2,0)
n[3,]<-c(0,-1)
print((m-n)[1:2,])
sum(abs(m-n))
5%%2
```
The following creates a distance table for hamming distance:
```{r}
hd_matrix <- matrix(,nrow=304,ncol=304)
for(i in 1:304){
  for(j in i:304){
    hd_matrix[i,j] <- hamming(gt_file[,i+1],gt_file[,j+1])
    hd_matrix[j,i] <- hd_matrix[i,j]
  }
  if (i%%20 == 0) print(i)
}

```


```{r}
#hd_matrix <- matrix()

cl_tree <- hclust(as.dist(hd_matrix),method = 'complete')
clusters <- cutree(cl_tree,h=5)
cent<- NULL
for (i in 1:5){
  cent<-rbind(cent, colMeans(hd_matrix[clusters == i]))
}

cl_cutree<-hclust(as.dist(cent),method = 'complete', members = table(clusters))
plot(clusters)

table_colors<-clusters[]
table_colors[table_colors == 1]<-'red'
table_colors[table_colors==2]<-'blue'
table_colors[table_colors == 3]<-'green'

plot(cl_tree,main='Hamming Distance Clusters', hang=-1,labels=FALSE,xlab = "Genome Keys",col= table_colors)
clusters
cl_tree
```


```{r}
#Save plot in file

png("Hamming_clusters.png",width=1800,height = 400)
plot(cl_tree,main='Hamming Distance Clusters',hang=-1)
dev.off()

```
We do the same for Cityblock distances.

```{r}
library(circlize)
cd_matrix <- matrix(,nrow=304,ncol=304)
for(i in 1:304){
  for(j in i:304){
    cd_matrix[i,j] <- cityblock(gt_file[,i+1],gt_file[,j+1])
    cd_matrix[j,i] <- cd_matrix[i,j]
  }
  if (i%%1 == 0) print(i)
}

ct_tree1 <-hclust(dist(cd_matrix,method = "manhattan")) %>% 
  as.dendrogram() %>% 
  set("labels_cex", 0.5) %>%
  set("labels_col",value = 1:4,k=8) %>%
  set("branches_k_color", value = 1:4, k = 8) %>%
  set("branches_lty", c(1,1,3,1,1,2)) %>%
  set("labels",colnames(vcf_gt_match)) 


  par(mar = rep(0,4))
  plot(ct_tree,main='Genomic clusters') 

```


 
 
```{r}
png("Cityblock_clusters.png",width=800,height = 3600)
ct_tree1 <-hclust(dist(cd_matrix,method = "manhattan")) %>% 
  as.dendrogram() %>% 
  set("labels_cex", 0.5) %>%
  set("labels_col",value = 2:4,k=6) %>%
  set("branches_k_color", value = 2:4, k = 6) %>%
  set("labels",colnames(vcf_gt_match)) %>%
  plot(main='Genomic clusters') 
dev.off()
```

Now, put them together

```{r}

ct_tree1 <-hclust(dist(hd_matrix,method = "manhattan")) %>% 
  as.dendrogram() %>% 
  set("labels_cex", 0.5) %>%
  set("labels_col",value = 1:4,k=8) %>%
  set("branches_k_color", value = 1:4, k = 8) %>%
  set("labels",colnames(vcf_gt_match)) 

ct_tree2 <-hclust(dist(cd_matrix,method = "manhattan")) %>% 
  as.dendrogram() %>% 
  set("labels_cex", 0.5) %>%
  set("labels_col",labels_col(ct_tree1)[order(labels(ct_tree2))]) %>%
  set("branches_k_color", value = 1:4, k = 8) %>%
  set("labels",colnames(vcf_gt_match)) 

par(mfrow=c(1,2))
ct_tree1 %>%  plot(main="Distance 1 Clusters")
ct_tree2 %>%  plot(main ="Distance 2 Clusters")



```

```{r}

trees=dendlist(ct_tree1,ct_tree2) %>%
tanglegram(common_subtrees_color_branches = TRUE)
```

```{r}
length(labels_col(ct_tree1)[labels_col(ct_tree1)==labels_col(ct_tree2)])
```

```{r}
dend15 <- c(1:5) %>% dist %>% hclust(method = "average") %>% as.dendrogram
dend15 <- dend15 %>% set("labels_to_char")
dend51 <- dend15 %>% set("labels", as.character(5:1)) %>% match_order_by_labels(dend15)
dends_15_51 <- dendlist(dend15, dend51)
dends_15_51
```

