---
title: "Accessing Genome Data"
output:
  pdf_document: default
  html_notebook: default
---

reading a vcf file: 

```{r}
library(vcfR)

genome_table=read.vcfR('imp_TERRA_RIL_SNP.vcf')
```

 accessing the fix, meta, and gt files: 

This gives the data for the reference DNA and puts it into a fll table: be careful with accessing the gt file since it is so large

```{r}
reference = genome_table@fix
View(reference)
gt_file = genome_table@gt
meta_file = genome_table@meta

```

The meta file gives you the meta data of the file: 

```{r}
meta_file = genome_table@meta
meta_file
```

To access a single genome (there are 304), access a column of the gt file (genome data starts at column 2). The genome is presented in the a|b format, where $a,b \in \{ 0,1 \}$.  The way to read this is comparing to the reference.  The reference gives you the column of reference or dominant alleles, but also a table of the alternate.  a sequence with 0|0 means that the entry has the dominant on both sides of the Helix, 0|1 and 1|0 means a dominant on one side, alternate on the other, and 1|1 means alterante alleles on both sides. To get the genome keys, simply take the 0th row. This gives you the "format" key, the rest of the keys, and then the "reference genome".

```{r}
table1 = gt_file[,2]
key_names = gt_file[0,]

```


