---
title: "match_vcf_genesys_48"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Extracting genesys data 
```{r, message=F, warning=F}
library(dplyr)
accessions <- read.csv("~/Team5/data/df_traits_nutrition.csv")

genesys_data <- accessions %>%
  mutate(acceNumb = gsub("PI ", "PI", acceNumb))
```

### Sort genesys_data by the PIs
```{r}
genesys_data_sorted <- genesys_data[order(genesys_data$acceNumb),]
```

## Extracting vcf data 
```{r, message=F, warning=F}
if (!require('vcfR')) install.packages('vcfR')
library(vcfR)
vcf_data <- read.vcfR(file="/data/terraref/sites/genomics/derived_data/ril/gbs/imp_TERRA_RIL_SNP.vcf")

vcf_fix = vcf_data@fix
vcf_gt = vcf_data@gt
vcf_meta = vcf_data@meta

colnames(vcf_gt)<-gsub(".MERGE", "",colnames(vcf_gt))
```

##  Find matching cultivar IDs
```{r}
match_vcf<-colnames(vcf_gt) %in% genesys_data_sorted$acceNumb
colnames(vcf_gt)[match_vcf]

match_genesys<- genesys_data_sorted$acceNumb  %in% colnames(vcf_gt)
genesys_data_sorted[match_genesys,]$acceNumb
```

### Extract 19 matches from genesys
```{r}
genesys_data_match <- genesys_data_sorted[match_genesys,]
```

### Extract 19 matches from vcf
```{r}
vcf_gt_match <- vcf_gt[,match_vcf]
```

### compute genetic distance

The following is an allele distance table which will define allele differences
```{r}
names <-c('0|0','0|1','1|0','1|1')
allele_dif_table <- matrix(c(0,1,1,2,1,0,0,1,1,0,0,1,2,1,1,0), nrow = 4,ncol = 4, dimnames = list(names, names))
```

### function turns strings '0|1'  into vectors c(0,1) etc.
```{r}
tonumericform <- function(a){
  vec1<-matrix(nrow = length(a),ncol =2)
  vec1[a=='0|0']<-c(0,0)
  vec1[a=='1|1']<-c(1,1)
  vec1[a=='0|1']<-c(0,1)
  vec1[a == '1|0']<-c(1,0)
  return(as.numeric(vec1)) 
} 
```

# function turns strings '0|1' '0|0'  into vector (0,1,0,0) etc.
# ```{r}
# tonumericform2 <- function(a){
#   vec1<-matrix(nrow = length(a),ncol =2)
#   vec1[a=='0|0']<-c(0,0)
#   vec1[a=='1|1']<-c(1,1)
#   vec1[a=='0|1']<-c(0,1)
#   vec1[a == '1|0']<-c(1,0)
#   return(vec1) 
#   } 
# ```


cityblock distance
```{r}
cityblock<- function(a1,a2){
  vec1<-tonumericform(a1)
  vec2<-tonumericform(a2)
  point_dif<-vec1-vec2
  dif<- sum(abs(vec1-vec2))
  return(dif)
}
```

distance vector
```{r}
d_vector <- vector()
for(i in 1:48){
    d_vector[i] <- cityblock(vcf_gt_match[,i],rep('0|0',70693))
}
```

distance matrix
```{r}
d_matrix <- matrix(,nrow = 48,ncol=48)
for(i in 1:48){
  for(j in 1:48){
    d_matrix[i,j] <- cityblock(vcf_gt_match[,i],vcf_gt_match[,j])
  }
}
heatmap(d_matrix, sym=TRUE)
```


### tree kmeans
```{r}
#if (!require('dendextend')) install.packages('dendextend'); library('dendextend')
cl_tree_kmeans <- kmeans(as.dist(d_matrix),centers=3)
#d1=color_branches(cl_tree,k=5)
#plot(cl_tree_kmeans)
```

### tree hclus
```{r}
#if (!require('dendextend')) install.packages('dendextend'); library('dendextend')
cl_tree <- hclust(as.dist(d_matrix))
#d1=color_branches(cl_tree,k=5)
plot(cl_tree)

#cutree(cl_tree, k = NULL, h = NULL)
clusters <- cutree(cl_tree, k = 4)
```

vcf2binary(vcf_gt_match[,1])
```{r}
vcf2binary <- function(x){
    a1 <- as.numeric(substr(vcf_gt_match[,1], 1,1))
    a2 <- as.numeric(substr(vcf_gt_match[,1], 3,3))
    append(a1,a2)
}
```

### add d_vector to genesys_data_match as last column 
```{r}
genesys_data_match[,17] <- d_vector
```

```{r}
genesys_matrix <- data.matrix(genesys_data_match[,6:16], rownames.force = NA)
rownames(genesys_matrix) <- genesys_data_match[,'V17']
heatmap(scale(genesys_matrix),Rowv=as.dendrogram(cl_tree))
```

```{r}
#Save picture:
png(file="heatmap_48_dist2.png",width=500,height = 500)
heatmap(scale(genesys_matrix),Rowv=as.dendrogram(cl_tree))
dev.off()
```

# ```{r}
# plot(genesys_data_match[,'V17'], genesys_data_match[,6  ],
#      xlab="dist from reference", ylab=colnames(genesys_data_match)[6],
#     cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5,
#     col=c(1:3)[cl_tree_kmeans$cluster],pch=20 )
# ```
     
Play with correlations:
```{r}
#plot(genesys_data_match[,'V17'], genesys_data_match[,7  ])
#png(file="correletions_48_colors.png",width=2000,height = 1000)
par(mfrow = c(3, 4))  # 3 rows and 4 columns
for (i in c(1:11)) {
    plot(genesys_data_match[,'V17'], genesys_data_match[,i+5  ],
    xlab="dist from reference", ylab=colnames(genesys_data_match)[i+5],
    cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5,
    col=clusters,pch=20,cex=2)}
  
# for (i in c(1:11)) {
#     plot(c(1:11), log2(genesys_data_match[,i+5  ]),
#     xlab="cultivar number (arbitrary)", ylab=colnames(genesys_data_match)[i+5],
#     cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5,
#     col=clusters,pch=20,cex=2)
  
#plot(genesys_data_match[,4], genesys_data_match[,5], col=clusters)
#dev.off()
                                          
```

